finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4DispatcherBase.scala

```diff
@@ -110,7 +111,11 @@ trait Netty4DispatcherBase[SendMsg <: Message, RecvMsg <: Message] {
       case Throw(e) if closed.get =>
         log.debug("[%s] dispatcher closed: %s", prefix, e)
         Future.exception(e)
-
+      // if we attempted to cast a Http2Frame as a H1 frame, log a message & kill the
+      // connection.
+      case Throw(e: ClassCastException) if e.getMessage == H1RequestError =>
```

Comment:
I'm worried about this being quite brittle but... I can't really see a better way

Reply:
Perhaps remove the message equality check and make the log more generic?  "HTTP/2 router could not handle non-HTTP/2 request!"?  WDYT?

Reply:
I considered that, but it seemed unwise to assume _all_ `ClassCastException`s were due to a mismatched request protocol. If a `ClassCastException` was thrown for some other reason, logging the protocol error message would obscure that failure...

What about a test that the message contains `"Transport.cast failed."`?

Reply:
:+1: